quickwit/quickwit-control-plane/src/scheduler.rs

```diff
@@ -17,37 +17,585 @@
+        let source_configs: HashMap<IndexSourceId, SourceConfig> = indexes_metadatas
+            .into_iter()
+            .flat_map(|index_metadata| {
+                index_metadata
+                    .sources
+                    .into_iter()
+                    .map(move |(source_id, source_config)| {
+                        (
+                            IndexSourceId {
+                                index_id: index_metadata.index_config.index_id.to_string(),
+                                source_id,
+                            },
+                            source_config,
+                        )
+                    })
+            })
+            .collect();
+        let indexing_tasks = build_indexing_plan(&indexers, &source_configs);
+        let new_physical_plan =
+            build_physical_indexing_plan(&indexers, &source_configs, indexing_tasks);
+        self.apply_physical_indexing_plan(&indexers, new_physical_plan)
+            .await;
+        Ok(())
+    }
+
+    /// Checks if the last applied plan corresponds to the running indexing tasks present in the
+    /// chitchat cluster state. If true, do nothing.
+    /// If false, schedule a new indexing plan.
+    async fn control_running_plan(&mut self) -> anyhow::Result<()> {
+        let last_applied_plan =
+            if let Some(last_applied_plan) = self.state.last_applied_physical_plan.as_ref() {
+                last_applied_plan
+            } else {
+                return Ok(());
+            };
+
+        // TODO(fmassot): once the actor framekwork can correctly simulate time advance, change
+        // `MIN_DURATION_BETWEEN_SCHEDULING_SECS` value for test and correclty test it. Currently, I
+        // put its value 0 to make test simple.
+        if let Some(last_applied_plan_timestamp) = self.state.last_applied_plan_timestamp.as_ref() {
+            if (OffsetDateTime::now_utc().unix_timestamp() - last_applied_plan_timestamp)
+                < MIN_DURATION_BETWEEN_SCHEDULING_SECS as i64
+            {
+                return Ok(());
+            }
+        }
+        let indexers = self.get_indexers_from_cluster_state().await;
+        let running_indexing_tasks_by_node_id: HashMap<String, Vec<IndexingTask>> = indexers
+            .into_iter()
+            .map(|cluster_member| {
+                (
+                    cluster_member.node_id,
+                    cluster_member
+                        .running_indexing_plan
+                        .map(|plan| plan.indexing_tasks)
+                        .unwrap_or_default(),
+                )
+            })
+            .collect();
+        if !are_indexing_plans_equal(
+            &running_indexing_tasks_by_node_id,
+            last_applied_plan.indexing_tasks_per_node(),
+        ) {
+            info!("Running plan and last applied plan differs: schedule an indexing plan.");
+            self.schedule_indexing_plan().await?;
```

Comment:
I wonder if reschedule is always the right choice here...

If the nodeset is unchanged, this is not the best way to go is it? 
Can we at least add a TODO and create an issue to revisit this logic?

Reply:
Indeed, we probably want just to reapply the same plan on the nodes where we don't have the same tasks.